scripts/copy_files_to_release.py

```diff
@@ -0,0 +1,122 @@
+
+# pylint: disable=E0401,E0611
+from cpg_utils.config import get_config
+from cpg_utils.cloud import get_path_components_from_gcp_path
+
+client = storage.Client()
+
+
+def check_paths_exist(paths: list[str]):
+    """
+    Checks a list of gs:// paths to see if they point to an existing blob
+    Logs the invalid paths if any are found
+    """
+    invalid_paths = False
+    for path in paths:
+        # gsutil ls <path> returns '<path>\n' if path exists
+        result = subprocess.run(
+            ['gsutil', 'ls', path], check=True, capture_output=True, text=True
+        ).stdout.strip('\n')
+        if result == path:
+            continue
+        # If path does not exist, log the path and set invalid_paths to True
+        logging.info(f'Invalid path: {path}')
+        invalid_paths = True
+
+    if invalid_paths:
+        return False
+    return True
+
+
+def copy_to_release(project: str, billing_project: str, paths: list[str]):
+    """
+    Copy many files from main bucket paths to the release bucket with todays date as directory
+    """
+    today = time.strftime('%Y-%m-%d')
+    release_path = f'gs://cpg-{project}-release/{today}/'
+
+    subprocess.run(
+        [
+            'gcloud',
+            'storage',
+            '--billing-project',
+            billing_project,
+            'cp',
+            *paths,
+            release_path,
+        ],
+        check=True,
+    )
+    logging.info(f'Copied {paths} into {release_path}')
+
+
+@click.command()
+@click.option('--project', '-p', help='Metamist name of the project', default='')
+@click.option('--billing-project', '-b', help='The GCP billing project to use')
+@click.argument('urls_file_path')
+def main(project: str, billing_project: str, urls_file_path: str):
+    """
+
+    Parameters
+    ----------
+    project :   a metamist dataset name, optional as it can be pulled from the AR config
+    billing_project :    a GCP project ID to bill to
+    urls_file_path :   a full GS path to a file containing the links to move into the release bucket
+    """
+    if not project:
+        config = get_config()
+        project = config['workflow']['dataset']
+
+    if not billing_project:
+        billing_project = project
+
+    if not urls_file_path.startswith(f'gs://'):
+        raise ValueError('url_file must be a fully qualified GS path')
+
+    path_components = get_path_components_from_gcp_path(urls_file_path)
+
+    bucket = path_components['bucket']
+    file_name = path_components['file']
+    if path_components['suffix']:
+        file_name = os.path.join(path_components['suffix'], path_components['file'])
+
+    input_bucket = client.get_bucket(bucket)
+    blob = input_bucket.get_blob(file_name)
+    if not blob:
+        raise RuntimeError(f'No file found at url_file path {urls_file_path}')
+
+    # Download and read the file containing the links to copy into release
+    blob.download_to_filename(file_name)
+    with open(file_name, 'r', encoding='ascii') as f:
```

Comment:
This will work fine, but if we know we're expecting a GCP file we can just read it using cloudpathlib.

```
from cpg_utils import to_path

with to_path(urls_file_path).open(encoding='utf-8') as f:
    paths = [line.rstrip() for line in f]
```

encoding-wise, so long as you're consistent with how it's written and read you'll be fine. I'm a fan of UTF-8 for unicode handling, but if you groom out any possible special characters you should be OK.

Reply:
This trims out a lot of the clutter. Many thanks @MattWellie!